File: Julia/juliaset.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <complex.h>

double julia(complex double z, complex double c, int max_iters);
double** makeGraph(int width, int height, double xmin, double xmax, double ymin, double ymax, complex double c, int max_iters);
void writeGraph(char* fname, double** graph, int width, int height);
complex double juliaFunc(complex double z);

int main() {
	int width = 1920;
	int height = 1080;
	double xmin = -1.5;
	double xmax = 1.5;
	double ymin = -1;
	double ymax = 1.0;
	complex double c = -0.621 + 0.128*I;
	int max_iters = 1000;
	double	**graph = makeGraph(width, height, xmin, xmax, ymin, ymax, c, max_iters);
	writeGraph("julia.txt", graph, width, height);
	free(graph);
	return 0;
}

complex double juliaFunc(complex double z) {
	return cexp(z*z*z);
}

double julia(complex double z, complex double c, int max_iters) {
	double i = 0;
	double eps = 0.00000001;
	complex double z0 = z;
	while (i < max_iters) {
		z = juliaFunc(z) + c;
		if ( creal(z)*creal(z) + cimag(z)*cimag(z) > 100000)
			return  log10(i/max_iters + eps);
		i++;	
	}
	return 0; 
}

double** makeGraph(int width, int height, double xmin, double xmax, double ymin, double ymax, complex double c, int max_iters) {
	double** graph = malloc(height*sizeof(double*));
	for (size_t y = 0; y < height; ++y) 
		graph[y] =  malloc( width*sizeof(double));
	double juliaVal;
	double zreal, zimag;
	complex double z;
	for (int y = 0; y < height; ++y) {
		for (int x = 0; x < width; ++x) {
			zreal = ((float)x)/width * (xmax - xmin) + xmin;
			zimag = ((float)y)/height * (ymax - ymin) + ymin;
			z = zreal + I*zimag;
			graph[y][x] = julia(z,c,max_iters);
		}
	}
	return graph;
}

void writeGraph(char* fname, double** graph, int width, int height) {
	FILE* fp;
	fp = fopen(fname, "w");

	for (int y = 0; y < height; ++y) {
		for (int x = 0; x < width; ++x)
			fprintf(fp, "%f\t", graph[y][x]);
		fprintf(fp, "\n");
	}
	fclose(fp);
}

File: Julia/juliaset.py
import cmath
from math import log10
import numpy as np
import matplotlib.pyplot as plt

def julia(z, c, max_iters):
	count = 0
	eps = 1e-8
	while count < max_iters:
		z = z*z + c
		if abs(z) > 2:
			return log10( float(count)/max_iters + eps)
		count += 1
	return 0

def makeGraph(width, height, xmin, xmax, ymin, ymax, c, max_iters):
	graph = np.zeros( (height,width) )
	for row in xrange(width):
		for col in xrange(height):
			zreal = float(row)/width * (xmax - xmin) + xmin
			zimag = float(col)/height * (ymax - ymin) + ymin
			z = zreal + zimag*1j
			graph[col][row] = julia(z, c, max_iters)
	return graph

def showGraph(graph, xmin, xmax, ymin, ymax): 
	plt.imshow(graph, cmap='hot', extent=[xmin,xmax,ymin,ymax])
	plt.show()
if __name__ == '__main__':
	xmin = -2
	xmax = 2
	ymin = -2
	ymax = 2
	max_iters = 256
	c = 0
	
	width = 1920
	height =  1080

	graph = makeGraph(width, height, xmin, xmax, ymin, ymax, c, max_iters)
	showGraph(graph, xmin, xmax, ymin, ymax)
